File: Practicas/RobocodeLabo-1.10.0/RobocodeLabo/src/laboratorio/NicoustinRobot.java
package laboratorio;

import estrategas.Estratega;
import estrategias.Estrategia;
import estrategias.StrategyMapper;
import estrategias.StrategyEvaluator;
import robocode.*;

/**
 * NicoustinRobot - Robot principal del laboratorio
 * 
 * Implementa un sistema de estrategias dinámico que se adapta según el número de enemigos:
 * - Inicio (muchos enemigos): EstrategiaWalls (supervivencia defensiva)
 * - Final (pocos enemigos): EstrategiaAntiWalls (caza ofensiva)
 * 
 * Utiliza el patrón Strategy Pattern con evaluación automática de situaciones.
 */
public class NicoustinRobot extends JuniorRobot {

    private Estratega estratega;
    private Estrategia estrategia;
    private StrategyEvaluator strategyEvaluator;
    private String currentSituationKey = "start";

    /**
     * Constructor predeterminado para Robocode
     * Inicializa con el evaluador y la estrategia apropiada según la situación actual
     */
    public NicoustinRobot() {
        // Configurar el evaluador de estrategias con referencia al robot
        this.strategyEvaluator = new StrategyEvaluator(this);
        
        // Inicializar con la estrategia apropiada
        this.estrategia = StrategyMapper.getStrategyForSituation(this);
        this.currentSituationKey = StrategyMapper.getCurrentSituation();
    }
    
    /**
     * Constructor con estrategia y evaluador personalizados (para testing)
     */
    public NicoustinRobot(Estrategia estrategia, StrategyEvaluator evaluator) {
        this.strategyEvaluator = evaluator;
        setEstrategia(estrategia);
    }

    /**
     * Establece una nueva estrategia y la vincula al robot
     */
    public void setEstrategia(Estrategia estrategia) {
        this.estrategia = estrategia;
        estrategia.setRobot(this);
    }
    
    /**
     * Establece un nuevo evaluador de estrategias
     */
    public void setStrategyEvaluator(StrategyEvaluator evaluator) {
        this.strategyEvaluator = evaluator;
    }
    
    /**
     * Evalúa la situación actual y cambia de estrategia si es necesario
     * Se ejecuta antes de cada acción importante
     */
    /// Tendriamos que cambiar este, despues lo cambiamos del todo
    private void updateStrategy() {
        if (estrategia != null && strategyEvaluator != null) {
            String newSituationKey = strategyEvaluator.evaluateStrategy();
            
            if (strategyEvaluator.shouldChangeStrategy(currentSituationKey, newSituationKey)) {
                System.out.println("🔄 NICOUSTIN: Cambiando estrategia '" + currentSituationKey + "' → '" + newSituationKey + "'");
                
                this.currentSituationKey = newSituationKey;
                StrategyMapper.updateSituation(newSituationKey);
                this.estrategia = StrategyMapper.getStrategyForSituation(this);
            }
        }
    }

    /**
     * private void updateStrategy(){
     *     if (estratega != null){
     *         Estrategia newStrat = estratega.chooseStrategy(this, currentSituationKey);
     *         if (newStrat != null && !newStrat.getClass().equals(estrategia.getClass())){
     *             System.out.println("🔄 NICOUSTIN: Cambiando estrategia '" + currentSituationKey + "' → '" + newSituationKey + "'");
     *             setEstrategia(newStrat);
     *         }
     *     }
     * }
     */

    /**
     * Y tendriamos tambien
     *
     * public void setEstratega(Estratega estratega){
     *     this.estratega = estratega;
     * }
     */

    /**
     * Bucle principal del robot
     */
    @Override	
    public void run() {
        updateStrategy();
        estrategia.runB(this);
    }

    /**
     * Evento: Robot enemigo detectado
     */
    public void onScannedRobot() {
        updateStrategy();
        estrategia.onScannedRobot();
    }

    /**
     * Evento: Impacto de bala recibido
     */
    @Override
    public void onHitByBullet() {
        updateStrategy();
        estrategia.onHitByBullet();
    }
    
    /**
     * Evento: Colisión con pared
     */
    @Override
    public void onHitWall() {
        updateStrategy();
        estrategia.onHitWall();
    }
}

File: Practicas/Backup/src/laboratorio/NicoustinRobot.kt
package laboratorio

import robocode.*
import estrategias.Estrategia
import estrategias.StrategyEvaluator
import estrategias.StrategyMapper

class NicoustinRobot() : JuniorRobot() {
    private var estrategia : Estrategia? = null
    private var stategyEvaluator : StrategyEvaluator? = null
    private var currentSituationKey : String = "start"

    init {
        stategyEvaluator = stategyEvaluator(this)
        estrategia = StrategyMapper.getStrategyForSituation(this)
        currentSituationKey = StrategyMapper.getCurrentSituation()
    }

    constructor(estrategia: Estrategia, evaluator: StrategyEvaluator) : this() {
        strategyEvaluator = evaluator
        setEstrategia(estrategia)
    }

    fun setEstrategia(estrategia: Estrategia) {
        this.estrategia = estrategia
        estrategia.setRobot(this)
    }

    fun setStrategyEvaluator(evaluator: StrategyEvaluator) {
        this.stategyEvaluator = evaluator
    }

    private fun updateStrategy(){
        if(estrategia != null && stategyEvaluator != null){
            val newSituationKey = stategyEvaluator!!.evaluateStrategy()

            if(stategyEvaluator!!.shouldChangeStrategy(currentSituationKey, newSituationKey)){
                println("NICOUSTIN: Cambiando estrategia '$currentSituationKey' -> '$newSituationKey'")

                currentSituationKey = newSituationKey
                StrategyMapper.updateSituation(newSituationKey)
                estrategia = StrategyMapper.getStrategyForSituation(this)
            }
        }
    }
    override fun run() {
        updateStrategy()
        estrategia?.runB(this)
    }

    /**
     * Evento: Robot enemigo detectado
     */
    fun onScannedRobot() {
        updateStrategy()
        estrategia?.onScannedRobot()
    }

    /**
     * Evento: Impacto de bala recibido
     */
    override fun onHitByBullet() {
        updateStrategy()
        estrategia?.onHitByBullet()
    }

    /**
     * Evento: Colisión con pared
     */
    override fun onHitWall() {
        updateStrategy()
        estrategia?.onHitWall()
    }
}